### PR TITLE
Expand Group2Seq

### PR DESCRIPTION
**Pull request by calebmkim**
Expands Group2Seq in the manner discussed in #1381. 

It produces a lot of unused assignments, so I can try to take #1384 (@priyasrikumar if you haven't already started). It's failing a test bc of unused assignments, so maybe even integrate it into this PR?

Also, going with this will allow us to get rid of #1381.

### CODE REVIEW

**Comment by rachitnigam** [comment]
I think a bunch of failures are coming from a combination of generating comb groups which don’t get used and the “ControlPorts” analysis panic over that

**Comment by priyasrikumar** [comment]
Hey! I've started on the passes, but if you want to take a stab at it/finish it off please feel free!!

File: calyx-opt/src/passes/dead_assignment_removal.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
+use itertools::Itertools;
+
+use crate::traversal::{Action, Named, VisResult, Visitor};
+use calyx_ir::{self as ir};
+use std::collections::{HashMap, HashSet};
+
+// maps combinational combinational components to set of all combinational components that it reads from
+// so the entries are (comb comp, <set of comb components that write to comb comp>)
+fn get_comb_depdendence_map(
```
Comment (rachitnigam) [inline calyx-opt/src/passes/dead_assignment_removal.rs:9]: typo in function name

File: calyx-opt/src/passes/dead_assignment_removal.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    non_comb_writes
+}
+
+/// Removes unused assigns from groups.
+/// Analyzes the writes to combinational cells in groups
+/// In order for a combinational cell to be considered "used", it must:
+/// 1) write to a non-combinational cell/group hole
+/// 2) write to a non-combinational cell that has been shown to be "used"
+#[derive(Default)]
+pub struct DeadAssignmentRemoval {}
```
Comment (rachitnigam) [inline calyx-opt/src/passes/dead_assignment_removal.rs:67]:
```suggestion
pub struct DeadAssignmentRemoval;
```

**Comment by rachitnigam** [comment]
LGTM! Pro-tip: you can automatically close PRs/Issues using Github's [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) and mentioning the PR/Issue in the top-comment or adding it to the "Development" bar on the right hand side of this page

**Pull request merged**